{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "80d684b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.21.0\n",
      "1.10.1\n",
      "1.24.2\n",
      "2.0.0\n",
      "9-20b_hifi.mm2-deepvar.t2tv2.indel.vcf.gz\n",
      "9-20b_hifi.mm2-sniffles.t2tv2.sv.vcf.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::idx_find_and_load] Could not retrieve index file for '/home/ebertp/work/data/raw_vcf/9-20b_hifi.mm2-sniffles.t2tv2.sv.vcf.gz'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    location filter_status call_type variant_type attribute      statistic   \n",
      "3     genome          PASS   PRECISE          INS    GT:1/1          count  \\\n",
      "5     genome          PASS   PRECISE          INS    GT:0/1          count   \n",
      "567   genome          PASS   PRECISE          INS    length         pct_01   \n",
      "568   genome          PASS   PRECISE          INS    length         pct_05   \n",
      "569   genome          PASS   PRECISE          INS    length      Q1_pct_25   \n",
      "570   genome          PASS   PRECISE          INS    length  median_pct_50   \n",
      "566   genome          PASS   PRECISE          INS    length           mean   \n",
      "571   genome          PASS   PRECISE          INS    length      Q3_pct_75   \n",
      "572   genome          PASS   PRECISE          INS    length         pct_95   \n",
      "573   genome          PASS   PRECISE          INS    length         pct_99   \n",
      "599   genome          PASS   PRECISE          INS   support         pct_01   \n",
      "600   genome          PASS   PRECISE          INS   support         pct_05   \n",
      "601   genome          PASS   PRECISE          INS   support      Q1_pct_25   \n",
      "602   genome          PASS   PRECISE          INS   support  median_pct_50   \n",
      "598   genome          PASS   PRECISE          INS   support           mean   \n",
      "603   genome          PASS   PRECISE          INS   support      Q3_pct_75   \n",
      "604   genome          PASS   PRECISE          INS   support         pct_95   \n",
      "605   genome          PASS   PRECISE          INS   support         pct_99   \n",
      "583   genome          PASS   PRECISE          INS   quality         pct_01   \n",
      "584   genome          PASS   PRECISE          INS   quality         pct_05   \n",
      "585   genome          PASS   PRECISE          INS   quality      Q1_pct_25   \n",
      "586   genome          PASS   PRECISE          INS   quality  median_pct_50   \n",
      "582   genome          PASS   PRECISE          INS   quality           mean   \n",
      "587   genome          PASS   PRECISE          INS   quality      Q3_pct_75   \n",
      "588   genome          PASS   PRECISE          INS   quality         pct_95   \n",
      "589   genome          PASS   PRECISE          INS   quality         pct_99   \n",
      "106   genome          PASS   PRECISE          INS    GT:./.          count   \n",
      "15    genome          PASS   PRECISE          DEL    GT:1/1          count   \n",
      "1     genome          PASS   PRECISE          DEL    GT:0/1          count   \n",
      "519   genome          PASS   PRECISE          DEL    length         pct_01   \n",
      "\n",
      "          value           sort_order  \n",
      "3     3705.0000   (0, 0, 0, 1, 1, 0)  \n",
      "5     9345.0000   (0, 0, 0, 1, 2, 0)  \n",
      "567     35.0000   (0, 0, 0, 1, 5, 1)  \n",
      "568     36.0000   (0, 0, 0, 1, 5, 5)  \n",
      "569     48.0000  (0, 0, 0, 1, 5, 25)  \n",
      "570     85.0000  (0, 0, 0, 1, 5, 50)  \n",
      "566    332.0000  (0, 0, 0, 1, 5, 55)  \n",
      "571    273.0000  (0, 0, 0, 1, 5, 75)  \n",
      "572   1390.0000  (0, 0, 0, 1, 5, 95)  \n",
      "573   4959.0000  (0, 0, 0, 1, 5, 99)  \n",
      "599      2.0000   (0, 0, 0, 1, 6, 1)  \n",
      "600      3.0000   (0, 0, 0, 1, 6, 5)  \n",
      "601      5.0000  (0, 0, 0, 1, 6, 25)  \n",
      "602      7.0000  (0, 0, 0, 1, 6, 50)  \n",
      "598      8.0000  (0, 0, 0, 1, 6, 55)  \n",
      "603     10.0000  (0, 0, 0, 1, 6, 75)  \n",
      "604     16.0000  (0, 0, 0, 1, 6, 95)  \n",
      "605     22.0000  (0, 0, 0, 1, 6, 99)  \n",
      "583     25.0000   (0, 0, 0, 1, 7, 1)  \n",
      "584     35.0000   (0, 0, 0, 1, 7, 5)  \n",
      "585     60.0000  (0, 0, 0, 1, 7, 25)  \n",
      "586     60.0000  (0, 0, 0, 1, 7, 50)  \n",
      "582     57.3043  (0, 0, 0, 1, 7, 55)  \n",
      "587     60.0000  (0, 0, 0, 1, 7, 75)  \n",
      "588     60.0000  (0, 0, 0, 1, 7, 95)  \n",
      "589     60.0000  (0, 0, 0, 1, 7, 99)  \n",
      "106     25.0000   (0, 0, 0, 1, 8, 0)  \n",
      "15    4754.0000   (0, 0, 0, 2, 1, 0)  \n",
      "1    10779.0000   (0, 0, 0, 2, 2, 0)  \n",
      "519     35.0000   (0, 0, 0, 2, 5, 1)  \n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "No active exception to reraise",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[62], line 250\u001b[0m\n\u001b[1;32m    248\u001b[0m \u001b[38;5;66;03m#df.drop(\"sort_order\", axis=1, inplace=True)\u001b[39;00m\n\u001b[1;32m    249\u001b[0m \u001b[38;5;28mprint\u001b[39m(df\u001b[38;5;241m.\u001b[39mhead(\u001b[38;5;241m30\u001b[39m))\n\u001b[0;32m--> 250\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m\n",
      "\u001b[0;31mRuntimeError\u001b[0m: No active exception to reraise"
     ]
    }
   ],
   "source": [
    "import pathlib as pl\n",
    "import pysam\n",
    "import collections as col\n",
    "import numpy as np\n",
    "import scipy\n",
    "import scipy.stats as scistats\n",
    "import pandas as pd\n",
    "import re as re\n",
    "\n",
    "print(pysam.__version__)\n",
    "print(scipy.__version__)\n",
    "print(np.__version__)\n",
    "print(pd.__version__)\n",
    "\n",
    "\n",
    "input_dir = pl.Path(\"/home/ebertp/work/data/raw_vcf\")\n",
    "\n",
    "vcfs = input_dir.glob(\"*.vcf.gz\")\n",
    "\n",
    "gt_map = lambda x: \".\" if x is None else x\n",
    "\n",
    "\n",
    "class StatisticsSorter():\n",
    "    \n",
    "    loc_order = {\n",
    "        \"wg\": 0,\n",
    "        \"genome\": 0,\n",
    "        \"chrX\": 23,\n",
    "        \"X\": 23,\n",
    "        \"chrY\": 24,\n",
    "        \"Y\": 24,\n",
    "        \"chrM\": 25,\n",
    "        \"chrMT\": 25,\n",
    "        \"M\": 25,\n",
    "        \"MT\": 25\n",
    "    }\n",
    "    \n",
    "    filter_order = {\n",
    "        \"PASS\": 0,\n",
    "        \"RefCall\": 1,\n",
    "        \"Refcall\": 1\n",
    "    }\n",
    "    \n",
    "    calltype_order = {\n",
    "        \"PRECISE\": 0,\n",
    "        \"IMPRECISE\": 1,\n",
    "        \"UNSPECIFIED\": 2,\n",
    "        \"UNKNOWN\": 3\n",
    "    }\n",
    "    \n",
    "    vartype_order = {\n",
    "        \"SNV\": 0,\n",
    "        \"INS\": 1,\n",
    "        \"DEL\": 2,\n",
    "        \"DUP\": 3,\n",
    "        \"INV\": 4,\n",
    "        \"BND\": 5\n",
    "    }\n",
    "    \n",
    "    attribute_order = {\n",
    "        \"GT:0/1\": 0,\n",
    "        \"GT:1/1\": 1,\n",
    "        \"GT:0/1\": 2,\n",
    "        \"GT:1/0\": 3,\n",
    "        \"GT:0/0\": 4,\n",
    "        \"length\": 5,\n",
    "        \"support\": 6,\n",
    "        \"quality\": 7\n",
    "    }\n",
    "    \n",
    "    statistics_order = {\n",
    "        \"count\": 0,\n",
    "        \"mean\": 55\n",
    "    }\n",
    "    \n",
    "    def __init__(self, locations, filters, calltypes, vartypes, attributes, statistics):\n",
    "        self = self._init_location_order(locations)\n",
    "        self = self._init_generic_order(filters, self.filter_order)\n",
    "        self = self._init_generic_order(calltypes, self.calltype_order)\n",
    "        self = self._init_generic_order(vartypes, self.vartype_order)\n",
    "        self = self._init_generic_order(attributes, self.attribute_order)\n",
    "        self = self._init_statistics_order(statistics)\n",
    "        return None\n",
    "    \n",
    "    def _init_location_order(self, locations):\n",
    "        \n",
    "        order_max = max(self.loc_order.values())\n",
    "        match_autosome = re.compile(\"^(chr)?[0-9]{1,2}$\")\n",
    "        for loc in sorted(locations):\n",
    "            mobj = match_autosome.search(loc)\n",
    "            if loc in self.loc_order:\n",
    "                continue\n",
    "            elif mobj is not None:\n",
    "                order_num = int(loc.strip(\"chr\"))\n",
    "                self.loc_order[loc] = order_num\n",
    "            else:\n",
    "                order_max += 1\n",
    "                self.loc_order[loc] = order_max\n",
    "        return self\n",
    "    \n",
    "    def _init_statistics_order(self, statistics):\n",
    "        \n",
    "        order_max = max(self.statistics_order.values())\n",
    "        for statistic in sorted(statistics):\n",
    "            if statistic in self.statistics_order:\n",
    "                continue\n",
    "            elif \"pct_\" in statistic:\n",
    "                pct_num = int(statistic.rsplit(\"_\", 1)[-1])\n",
    "                self.statistics_order[statistic] = pct_num\n",
    "                order_max = max(pct_num, order_max)\n",
    "            else:\n",
    "                order_max += 1\n",
    "                self.statistics_order[statistic] = order_max\n",
    "        return self\n",
    "    \n",
    "    def _init_generic_order(self, values, value_map):\n",
    "        \n",
    "        order_max = max(value_map.values())\n",
    "        for value in sorted(values):\n",
    "            if value in value_map:\n",
    "                continue\n",
    "            else:\n",
    "                order_max += 1\n",
    "                value_map[value] = order_max\n",
    "        return self   \n",
    "                    \n",
    "    def get_order_number(self, data_row):\n",
    "        \n",
    "        order_loc = self.loc_order[data_row[\"location\"]]\n",
    "        order_filter = self.filter_order[data_row[\"filter_status\"]]\n",
    "        order_calltype = self.calltype_order[data_row[\"call_type\"]]\n",
    "        order_vartype = self.vartype_order[data_row[\"variant_type\"]]\n",
    "        order_attribute = self.attribute_order[data_row[\"attribute\"]]\n",
    "        order_statistic = self.statistics_order[data_row[\"statistic\"]]\n",
    "        order_num = order_loc, \\\n",
    "            order_filter, \\\n",
    "            order_calltype, \\\n",
    "            order_vartype, \\\n",
    "            order_attribute, \\\n",
    "            order_statistic\n",
    "        return order_num\n",
    "\n",
    "\n",
    "count_stats = col.Counter()\n",
    "agg_stats = col.defaultdict(list)\n",
    "for vcf_file in vcfs:\n",
    "    print(vcf_file.name)\n",
    "    fixed_var_type = vcf_file.name.rsplit(\".\", 3)[-3]\n",
    "    if fixed_var_type != \"sv\":\n",
    "        continue\n",
    "    with pysam.VariantFile(vcf_file) as vcf:\n",
    "        for record in vcf.fetch():\n",
    "            contig = record.contig\n",
    "            filter_category = \"|\".join(record.filter.keys())\n",
    "            if \"PRECISE\" in record.info.keys():\n",
    "                calltype = \"PRECISE\"\n",
    "            elif \"IMPRECISE\" in record.info.keys():\n",
    "                calltype = \"IMPRECISE\"\n",
    "            else:\n",
    "                calltype = \"UNSPECIFIED\"\n",
    "            if \"SUPPORT\" in record.info.keys():\n",
    "                read_support = record.info[\"SUPPORT\"]\n",
    "            elif \"RE\" in record.info.keys():\n",
    "                read_support = record.info[\"RE\"]\n",
    "            elif \"RNAMES\" in record.info.keys():\n",
    "                read_support = len(record.info[\"RNAMES\"])\n",
    "            else:\n",
    "                read_support = -1\n",
    "            try:\n",
    "                vartype = record.info[\"SVTYPE\"]\n",
    "            except KeyError:\n",
    "                vartype = fixed_var_type\n",
    "            try:\n",
    "                varlen = abs(record.info[\"SVLEN\"])\n",
    "            except KeyError:\n",
    "                ref_length = len(record.ref)\n",
    "                assert ref_length > 0\n",
    "                diff_lengths = set(abs(ref_length - len(alt)) for alt in record.alts)\n",
    "                if len(diff_lengths) == 1:\n",
    "                    varlen = max(1, diff_lengths.pop())\n",
    "                else:\n",
    "                    varlen = -1\n",
    "            quality = record.qual\n",
    "            vcf_samples = list(record.samples.keys())\n",
    "            assert len(vcf_samples) == 1\n",
    "            vcf_sample_name = vcf_samples[0]\n",
    "            sample_info = dict(record.samples[vcf_sample_name].items())\n",
    "            gt = sample_info[\"GT\"]\n",
    "            if read_support == -1:\n",
    "                dp = sample_info[\"DP\"]\n",
    "                read_support = dp\n",
    "            \n",
    "            genotype = f\"GT:{gt_map(gt[0])}/{gt_map(gt[1])}\"\n",
    "            count_stats[(contig, filter_category, calltype, vartype, genotype)] += 1\n",
    "            count_stats[(\"genome\", filter_category, calltype, vartype, genotype)] += 1\n",
    "            agg_stats[(contig, filter_category, calltype, vartype, \"length\")].append(varlen)\n",
    "            agg_stats[(\"genome\", filter_category, calltype, vartype, \"length\")].append(varlen)\n",
    "            agg_stats[(contig, filter_category, calltype, vartype, \"quality\")].append(quality)\n",
    "            agg_stats[(\"genome\", filter_category, calltype, vartype, \"quality\")].append(quality)\n",
    "            agg_stats[(contig, filter_category, calltype, vartype, \"support\")].append(read_support)\n",
    "            agg_stats[(\"genome\", filter_category, calltype, vartype, \"support\")].append(read_support)\n",
    "    \n",
    "    summary = []\n",
    "    for key, value in count_stats.items():\n",
    "        row = list(key)\n",
    "        row.extend([\"count\", value])\n",
    "        summary.append(row)\n",
    "    \n",
    "    for key, values in agg_stats.items():\n",
    "        row = list(key)\n",
    "        data_array = np.array(values)\n",
    "        pct_scores = scistats.scoreatpercentile(\n",
    "            data_array,\n",
    "            per=[1, 5, 25, 50, 75, 95, 99]\n",
    "        )\n",
    "        mean = data_array.mean()\n",
    "        if row[-1] in [\"length\", \"support\"]:\n",
    "            normalize = lambda x: int(round(x, 0))\n",
    "        else:\n",
    "            normalize = lambda x: float(round(x, 4))\n",
    "        labels = [\n",
    "            \"mean\", \"pct_01\", \"pct_05\",\n",
    "            \"Q1_pct_25\", \"median_pct_50\",\n",
    "            \"Q3_pct_75\", \"pct_95\", \"pct_99\"]\n",
    "        values = [mean] + list(pct_scores)\n",
    "        for l, v in zip(labels, values):\n",
    "            norm_v = normalize(v)\n",
    "            new_row = row + [l, norm_v]\n",
    "            summary.append(new_row)\n",
    "            \n",
    "    df = pd.DataFrame.from_records(\n",
    "        summary,\n",
    "        columns=[\n",
    "            \"location\", \"filter_status\", \"call_type\", \"variant_type\", \"attribute\", \"statistic\", \"value\"\n",
    "        ]\n",
    "    )\n",
    "    sorter = StatisticsSorter(\n",
    "        df[\"location\"].unique(),\n",
    "        df[\"filter_status\"].unique(),\n",
    "        df[\"call_type\"].unique(),\n",
    "        df[\"variant_type\"].unique(),\n",
    "        df[\"attribute\"].unique(),\n",
    "        df[\"statistic\"].unique()\n",
    "    )\n",
    "    \n",
    "    df[\"sort_order\"] = df.apply(sorter.get_order_number, axis=1)\n",
    "    df.sort_values(\"sort_order\", ascending=True, inplace=True)\n",
    "    #df.drop(\"sort_order\", axis=1, inplace=True)\n",
    "    print(df.head(30))\n",
    "    raise"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
